File: src/easy-111/recursion.ts
import { BinaryTreeNode } from 'data-structures/binary-tree'
import { isDef } from 'src/shared'

export function minDepth<V>(node: BinaryTreeNode<V> | null): number {
  if (!isDef(node)) return 0
  if (!isDef(node.left) && !isDef(node.right)) return 1

  let answer = Number.MAX_VALUE // or Number.MAX_SAFE_INTEGER
  if (isDef(node.left)) {
    answer = Math.min(minDepth(node.left), answer)
  }
  if (isDef(node.right)) {
    answer = Math.min(minDepth(node.right), answer)
  }
  return answer + 1
}

File: src/easy-111/recursion.go
package easy111

import "math"

func recursiveMinDepth(root *node) int {
	if root == nil {
		return 0
	}
	if root.Left == nil && root.Right == nil {
		return 1
	}
	depth := math.MaxInt32
	if root.Left != nil {
		depth = min(recursiveMinDepth(root.Left), depth)
	}
	if root.Right != nil {
		depth = min(recursiveMinDepth(root.Right), depth)
	}
	return depth + 1
}

func min(a, b int) int {
	if a > b {
		return b
	}
	return a
}
